lib/cf_deployer/driver/auto_scaling_group.rb

```diff
@@ -62,21 +62,42 @@ def desired_capacity_reached?
         healthy_instance_count >= desired_capacity
       end
 
-      private
+      def healthy_instance_ids
+        instances = auto_scaling_instances.select do |instance|
+          'HEALTHY'.casecmp(instance.health_status) == 0
+        end
+        instances.map(&:id)
+      end
+
+      def in_service_instance_ids
+        elbs = load_balancers
+        return [] if elbs.empty?
+
+        ids = elbs.collect(&:instances)
+                  .collect(&:health)
+                  .to_a
+                  .collect { |elb_healths|
+                     elb_healths.select { |health| health[:state] == 'InService' }
+                                .map { |health| health[:instance].id }
+                  }
+
+        ids.inject(:&)
+      end
 
       def healthy_instance_count
-        begin
-          count = auto_scaling_instances.count do |instance|
-            instance.health_status == 'HEALTHY' && (load_balancers.empty? || instance_in_service?( instance.ec2_instance ))
-          end
-          Log.info "Healthy instance count: #{count}"
-          count
-        rescue => e
-          Log.info "Unable to determine healthy instance count due to error: #{e.message}"
-          -1
+        AWS.memoize do
+          instances = healthy_instance_ids
+          instances &= in_service_instance_ids unless load_balancers.empty?
```

Comment:
As cf_deployer was originally written, no, each stack was not expected to have it's own ELB.

There are two Blue/Green deployment strategies.  One strategy is CnameSwap, which creates separate ELBs, and handles a swap by attaching a CNAME record to one or the other ELB.  The second strategy is AutoScalingGroupSwap (the one being modified here), which uses any number (zero or more) of fixed ELBs, and handles a deployment by scaling one or the other ASG up/down from 0. 

It sounds like a reasonable assumption (that the ASG healthcheck would pass before it was added to the ELB) - but I'm not sure that I can answer that question with certainty.  And it still does not account for the intended scenario of a stack with no ELB - which is explicitly called out in the test case referenced below.  I agree it would be simpler, but it would be a non-backwards-compatible change of functionality, that I think would be better addressed in a separate issue/release.  

CnameSwap - https://github.com/manheim/cf_deployer/blob/master/lib/cf_deployer/deployment_strategy/cname_swap.rb
AutoScalingGroupSwap - https://github.com/manheim/cf_deployer/blob/master/lib/cf_deployer/deployment_strategy/auto_scaling_group_swap.rb